### ISSUE TITLE
How to get `getElementError` function in Custom Query 

### ISSUE THREAD

**Issue by Yama-Tomo** [testing-library/dom-testing-library#508]
- @testing-library/dom@7.1.1

I have created custom query using this document.
https://testing-library.com/docs/dom-testing-library/api-helpers#custom-queries
But not found function getElementError in the queryHelpers.

I found the deletion of the declaration in #452.
Is the code below correct to do this?

```javascript
// Hmmm.. `getConfig` is not exported at the entry point.
const config = require('@testing-library/dom/dist/config')

export function getMyCustomElement(container) {
  ... snip ...
  if (!element) {
    throw config.getConfig().getElementError(`Unable to find an element....`, container)
  }
  ... snip ...
}
````

**Comment by eps1lon** [comment]
We have a dedicated section in our docs that explains how to build a custom error: https://testing-library.com/docs/dom-testing-library/api-helpers#custom-queries.

Hope that helps. If you have more questions I would suggest our [spectrum.chat](https://spectrum.chat/testing-library?tab=posts)

**Comment by Yama-Tomo** [comment]
@eps1lon My point is that it doesn't work according to the document.
Should I still ask the question there?

**Comment by eps1lon** [comment]
@Yama-Tomo The few bits you provided do not the documentation. You're importing from `dist` and using config instead of query helpers. For meaningful help we would need a complete example.

**Comment by Yama-Tomo** [comment]
@eps1lon I didn't write the origin.  I'm sorry.

I wrote a custom query that looks like the following.
(The actual custom query is more complicated, but we want to mention `getElementError`, so we've omitted it.

```typescript
import * as React from 'react';
import { render, queryHelpers } from '@testing-library/react';

const getMyCustomElement = (container: HTMLElement) => {
  throw queryHelpers.getElementError('throw error test', container);
};

it('example', () => {
  const { container } = render(<div />);
  getMyCustomElement(container);
});
````

There was an error when I ran the above code
`TypeError: _react2.queryHelpers.getElementError is not a function`

Next, I tried to dump `queryHelpers`

```diff
const getMyCustomElement = (container: HTMLElement) => {
+ console.log(queryHelpers);
  throw queryHelpers.getElementError('throw error test', container);
};
```

`getElementError` wasn't found.

```
    {
      getMultipleElementsFoundError: [Function: getMultipleElementsFoundError],
      queryAllByAttribute: [Function: queryAllByAttribute],
      queryByAttribute: [Function: queryByAttribute],
      makeSingleQuery: [Function: makeSingleQuery],
      makeGetAllQuery: [Function: makeGetAllQuery],
      makeFindQuery: [Function: makeFindQuery],
      buildQueries: [Function: buildQueries]
    }
```

**Comment by eps1lon** [comment]
What version of `@testing-library/react` and `@testing-library/dom` do you have installed? You're likely using an older version

**Comment by Yama-Tomo** [comment]
> - @testing-library/dom@7.1.1

@testing-library/react  is 10.0.1

I have upgraded testing-library/dom to 7.1.3 and I get the same error.

**Comment by eps1lon** [comment]
Confirmed in https://codesandbox.io/s/morning-pond-c1mps.

I was confused by the linked guide. I usually use `buildQuerie` which works fine. In the future please provide a full reproduction.

**Comment by Yama-Tomo** [comment]
>  I usually use buildQuerie which works fine.

The functions called by `buildQueries` gets `getElementError` via `getConfig`, so there is no error.

https://github.com/testing-library/dom-testing-library/blob/c0a4eb2c301fcf11a786a18b3008f7cad3a14692/src/query-helpers.js#L6
https://github.com/testing-library/dom-testing-library/blob/c0a4eb2c301fcf11a786a18b3008f7cad3a14692/src/query-helpers.js#L58

**Comment by eps1lon** [comment]
To be honest I don't know why we would expose this anyway. `buildQueries` seems like a much better helper. https://testing-library.com/docs/dom-testing-library/api-helpers#custom-queries not only is way toz convoluted for changing the attribute of `ByTestId` (uses `.bind`, can be configured now) but also is better suited for implementation by `buildQueries`.

It seems like we can safely remove that section or should find a better use case for a read of `getElementError`. Any thoughts?

/cc @testing-library/dom

**Comment by Yama-Tomo** [comment]
> buildQueries seems like a much better helper.

That's certainly true... 😄

For example, I think buildQueries is slightly excessive in the case of a simple case to get a single element.
In that case, it would be useful to be able to call getElementError directly.
```javascript
const getMyElement = (container) => {
  const selector = '.some > .selector'
  const element = container.querySelector(selector)
  if (!element) {
    throw queryHelpers.getElementError(`Unable to find an element with selector of: ${selector}`, container)
  }
  return element
}
```

_patterns using `buildQueries`_

```javascript
// raw_selector_queries.js
const [queryBy, getAll, getBy, findAll, findBy] = buildQueries(
  (container, ...rest) => container.querySelectorAll(...rest),
  (_, selector) => `Found multiple elements with selector of: ${selector}`,
  (_, selector) => `Unable to find an element with selector of: ${selector}`,
)

export { queryBy, getAll, getBy, findAll, findBy }

//------------------------------
import { getBy } from './raw_selector_queries'

const getMyElement = (container) => getBy(container, '.some > .selector')
````
However, if you code is large, a pattern using buildQueries may be preferable as it is DRY.



I don't know why the function that generates the error message should be private, but I will follow your team thoughts! ( it might be better to un export `getMultipleElementsFoundError` too.

**Comment by alexkrolick** [comment]
buildQueries is the way to go if you want to generate the full spectrum of query types. Building all of them means fewer surprises for your team because the usage is consistent with whatever they read in the docs. Prior to buildQueries being introduced all the queries were manually written and a few helpers may have been exposed that were meant to make it easier to match the builtins when writing your own.

**Comment by eps1lon** [comment]
> For example, I think buildQueries is slightly excessive in the case of a simple case to get a single element.

This seems like a good thing to me. It's not intended that you build custom queries for one-off cases. It's already hard enough to decide on the correct query to use. If you scale this up too much it'll get out of hand.

**Comment by eps1lon** [comment]
@kentcdodds I see you merged #510. Any opinion on https://github.com/testing-library/dom-testing-library/issues/508#issuecomment-605693421?

**Comment by kentcdodds** [comment]
Sorry I didn't respond. I honestly merged that because I didn't see anything harmful about it and forgot about the conversation going on here. I apologise.

At this point, I think we should just keep it as-is and just not recommend people use it for most cases.
